ocaml/xapi/xapi_vm.ml

```diff
@@ -1444,6 +1444,21 @@ let set_appliance ~__context ~self ~value =
   (* Update the VM's allowed operations - this will update the new appliance's operations, if valid. *)
   update_allowed_operations ~__context ~self
 
+let set_groups ~__context ~self ~value =
+  let anti_affinity_group_cnt =
+    value
+    |> List.filter (fun group ->
+           Db.VM_group.get_placement ~__context ~self:group = `anti_affinity
+       )
+    |> List.length
+  in
+  if anti_affinity_group_cnt > 1 then
+    raise
+      (Api_errors.Server_error
+         (Api_errors.vm_can_only_belong_to_one_anti_affinity_group, [])
+      ) ;
+  Db.VM.set_groups ~__context ~self ~value
```

Comment:
How about the following? In this way, it would be easier to add more assertions in the future.
```
let assert_only_one_antiaffinity value =
  value
  |> List.filter
       (fun group ->
         match Db.VM_group.get_placement ~__context ~self:group with
         | `anti_affinity ->
             true
         | _ ->
             false
       )
  |> function
  | _::_::_ ->
       raise
        (Api_errors.Server_error
           (Api_errors.vm_can_only_belong_to_one_anti_affinity_group, [])
        ) ;
  | _ ->
      ()
in
assert_only_one_antiaffinity value ;
Db.VM.set_groups ~__context ~self ~value
```